File: src/test/java/com/zepben/vertxutils/json/LazyJsonObjectTest.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.vertxutils.json;

import io.vertx.core.json.JsonObject;
import org.apache.commons.lang3.mutable.MutableInt;
import org.junit.jupiter.api.Test;

import static org.hamcrest.CoreMatchers.*;
import static org.hamcrest.MatcherAssert.assertThat;

public class LazyJsonObjectTest {

    private static final String KEY1 = "pi";
    private static final String KEY2 = "zepben";
    private static final double VALUE1 = Math.PI;
    private static final String VALUE2 = "Zeppelin Bend";

    @Test
    public void lazyObjectTest() {
        LazyJsonObject ljo = new LazyJsonObject(() -> {
            JsonObject jsonObject = new JsonObject();
            jsonObject.put(KEY1, VALUE1);
            jsonObject.put(KEY2, VALUE2);
            return jsonObject;
        });

        String encoded = ljo.encode();
        JsonObject decoded = new JsonObject(encoded);

        assertThat(decoded.getDouble(KEY1), equalTo(VALUE1));
        assertThat(decoded.getString(KEY2), equalTo(VALUE2));
    }

    @Test
    public void lazyIncludedMemberTest() {
        final MutableInt callCount = new MutableInt(0);

        LazyJsonObject jsonObject = new LazyJsonObject();
        jsonObject.put(KEY1, () -> {
            callCount.increment();
            return VALUE1;
        });
        jsonObject.put(KEY2, VALUE2);

        String encoded = jsonObject.encode();
        JsonObject decoded = new JsonObject(encoded);

        assertThat(decoded.getDouble(KEY1), equalTo(VALUE1));
        assertThat(decoded.getString(KEY2), equalTo(VALUE2));
        assertThat(callCount.getValue(), equalTo(1));
    }

    @Test
    public void lazyExcludedMemberTest() {
        final MutableInt callCount = new MutableInt(0);

        LazyJsonObject jsonObject = new LazyJsonObject();
        jsonObject.put(KEY1, () -> {
            callCount.increment();
            return VALUE1;
        });
        jsonObject.put(KEY2, VALUE2);

        jsonObject.remove("pi");

        String encoded = jsonObject.encode();
        JsonObject decoded = new JsonObject(encoded);

        assertThat(decoded.getDouble(KEY1), is(nullValue()));
        assertThat(decoded.getString(KEY2), equalTo(VALUE2));
        assertThat(callCount.getValue(), equalTo(0));
    }

    @Test
    public void lazyObjectWithIncludedLazyMemberTest() {
        final MutableInt callCount = new MutableInt(0);

        LazyJsonObject ljo = new LazyJsonObject(() -> {
            LazyJsonObject jsonObject = new LazyJsonObject();
            jsonObject.put(KEY1, () -> {
                callCount.increment();
                return VALUE1;
            });
            jsonObject.put(KEY2, VALUE2);
            return jsonObject;
        });

        String encoded = ljo.encode();
        JsonObject decoded = new JsonObject(encoded);

        assertThat(decoded.getDouble(KEY1), equalTo(VALUE1));
        assertThat(decoded.getString(KEY2), equalTo(VALUE2));

        assertThat(callCount.getValue(), equalTo(1));
    }

    @Test
    public void lazyObjectWithExcludedLazyMemberTest() {
        final MutableInt callCount = new MutableInt(0);

        LazyJsonObject ljo = new LazyJsonObject(() -> {
            LazyJsonObject jsonObject = new LazyJsonObject();
            jsonObject.put(KEY1, () -> {
                callCount.increment();
                return VALUE1;
            });
            jsonObject.put(KEY2, VALUE2);
            return jsonObject;
        });

        ljo.remove("pi");

        String encoded = ljo.encode();

        JsonObject decoded = new JsonObject(encoded);

        assertThat(decoded.getDouble(KEY1), is(nullValue()));
        assertThat(decoded.getString(KEY2), equalTo(VALUE2));

        assertThat(callCount.getValue(), equalTo(0));
    }

}

File: src/test/java/com/zepben/vertxutils/json/LazyJsonObjectTest.kt
/*
 * Copyright 2026 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.vertxutils.json

import com.zepben.testutils.junit.SystemLogExtension
import io.vertx.core.json.JsonObject
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.hamcrest.Matchers.nullValue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.RegisterExtension

class LazyJsonObjectTest {

    companion object {

        @JvmField
        @RegisterExtension
        val systemOut: SystemLogExtension = SystemLogExtension.SYSTEM_OUT.captureLog().muteOnSuccess()

        private const val KEY1 = "pi"
        private const val KEY2 = "zepben"
        private const val VALUE1 = Math.PI
        private const val VALUE2 = "Zeppelin Bend"
    }

    @Test
    fun lazyObjectTest() {
        val ljo = LazyJsonObject {
            val jsonObject = JsonObject()
            jsonObject.put(KEY1, VALUE1)
            jsonObject.put(KEY2, VALUE2)
            jsonObject
        }

        val encoded = ljo.encode()
        val decoded = JsonObject(encoded)

        assertThat(decoded.getDouble(KEY1), equalTo(VALUE1))
        assertThat(decoded.getString(KEY2), equalTo(VALUE2))
    }

    @Test
    fun lazyIncludedMemberTest() {
        var callCount = 0

        val jsonObject = LazyJsonObject()
        jsonObject.put(KEY1) {
            ++callCount
            VALUE1
        }
        jsonObject.put(KEY2, VALUE2)

        val encoded = jsonObject.encode()
        val decoded = JsonObject(encoded)

        assertThat(decoded.getDouble(KEY1), equalTo(VALUE1))
        assertThat(decoded.getString(KEY2), equalTo(VALUE2))
        assertThat(callCount, equalTo(1))
    }

    @Test
    fun lazyExcludedMemberTest() {
        var callCount = 0

        val jsonObject = LazyJsonObject()
        jsonObject.put(KEY1) {
            ++callCount
            VALUE1
        }
        jsonObject.put(KEY2, VALUE2)

        jsonObject.remove(KEY1)

        val encoded = jsonObject.encode()
        val decoded = JsonObject(encoded)

        assertThat(decoded.getDouble(KEY1), nullValue())
        assertThat(decoded.getString(KEY2), equalTo(VALUE2))
        assertThat(callCount, equalTo(0))
    }

    @Test
    fun lazyObjectWithIncludedLazyMemberTest() {
        var callCount = 0

        val ljo = LazyJsonObject {
            val jsonObject = LazyJsonObject()
            jsonObject.put(KEY1) {
                ++callCount
                VALUE1
            }
            jsonObject.put(KEY2, VALUE2)
            jsonObject
        }

        val encoded = ljo.encode()
        val decoded = JsonObject(encoded)

        assertThat(decoded.getDouble(KEY1), equalTo(VALUE1))
        assertThat(decoded.getString(KEY2), equalTo(VALUE2))

        assertThat(callCount, equalTo(1))
    }

    @Test
    fun lazyObjectWithExcludedLazyMemberTest() {
        var callCount = 0

        val ljo = LazyJsonObject {
            val jsonObject = LazyJsonObject()
            jsonObject.put(KEY1) {
                ++callCount
                VALUE1
            }
            jsonObject.put(KEY2, VALUE2)
            jsonObject
        }

        ljo.remove(KEY1)

        val encoded = ljo.encode()

        val decoded = JsonObject(encoded)

        assertThat(decoded.getDouble(KEY1), nullValue())
        assertThat(decoded.getString(KEY2), equalTo(VALUE2))

        assertThat(callCount, equalTo(0))
    }

}
